lib/ASLValue.ml

```diff
@@ -2,79 +2,119 @@ module ASLConstant = SymbConstant.Make (ASLScalar) (PteVal.No) (ASLBase.Instr)
     | ToBV -> "ToBV"
 
-  let do_op1 =
-    let ( let* ) = Option.bind in
-    let return_concrete s = Some (Constant.Concrete s) in
-    let as_concrete = function Constant.Concrete v -> Some v | _ -> None in
-    let as_concrete_vector = function
-      | Constant.ConcreteVector v -> Some v
-      | _ -> None
+  let ( let* ) = Option.bind
+  let return_concrete s = Some (Constant.Concrete s)
+  let as_concrete = function Constant.Concrete v -> Some v | _ -> None
+
+  let as_concrete_vector = function
+    | Constant.ConcreteVector v -> Some v
+    | _ -> None
+
+  let all_64_bits_positions = List.init 64 (( - ) 63)
+
+  let list_set =
+    let rec list_set acc n elt = function
+      | [] -> None
+      | _ :: t when n == 0 -> Some (List.rev acc @ (elt :: t))
+      | h :: t -> list_set (h :: acc) (n - 1) elt t
     in
-    let all_64_bits_positions = List.init 64 (( - ) 63) in
-    fun op cst ->
-      match op with
-      | Get i ->
-          let* vec = as_concrete_vector cst in
-          List.nth_opt vec i
-      | Set (i, elt) ->
-          let* vec = as_concrete_vector cst in
-          let* vec' = list_set i elt vec in
-          Some (Constant.ConcreteVector vec')
-      | ToInt -> (
-          match cst with
-          | Constant.Concrete s -> ASLScalar.convert_to_int s |> return_concrete
-          | Constant.Symbolic _ -> Some cst
-          | _ -> None)
-      | ToBV -> (
-          match cst with
-          | Constant.Concrete s -> ASLScalar.convert_to_bv s |> return_concrete
-          | Constant.Symbolic _ -> Some cst
-          | _ -> None)
-      | ToBool ->
-          let* s = as_concrete cst in
-          return_concrete (ASLScalar.convert_to_bool s)
-      | BVSlice positions -> (
-          match cst with
-          | Constant.Concrete s ->
-              let* s' = ASLScalar.try_extract_slice s positions in
-              return_concrete s'
-          | Constant.Symbolic x ->
-              if Misc.list_eq ( = ) positions all_64_bits_positions then
-                Some (Constant.Symbolic x)
-              else None
-          | _ -> None)
+    list_set []
+
+  let do_op op c1 c2 =
+    match op with
+    | Set i ->
+        let* vec = as_concrete_vector c1 in
+        let* vec' = list_set i c2 vec in
+        Some (Constant.ConcreteVector vec')
+    | Concat ->
+        let* s1 = as_concrete c1 in
+        let* s2 = as_concrete c2 in
+        let* s = ASLScalar.try_concat s1 s2 in
+        return_concrete s
+    | BVSliceSet positions ->
+        let* s1 = as_concrete c1 in
+        let* s2 = as_concrete c2 in
+        let* s = ASLScalar.try_write_slice positions s1 s2 in
+        return_concrete s
+
+  let do_op1 op cst =
+    match op with
+    | Get i ->
+        let* vec = as_concrete_vector cst in
+        List.nth_opt vec i
+    | ToIntS -> (
+        match cst with
+        | Constant.Concrete s ->
+            ASLScalar.convert_to_int_signed s |> return_concrete
+        | Constant.Symbolic _ -> Some cst
```

Comment:
It looks as if all the operators behave as the identity on "rigid" symbolic values. Probably worth an explanatory comment.

Reply:
Done that in cac5b6aa5659f8ccdb3597f39b96c99099c4ea1e